kernelci/data/kernelci_api.py

```diff
@@ -83,6 +95,47 @@ def get_nodes_by_commit_hash(self, commit_hash):
     def get_node_from_event(self, event):
         return self.get_node(event.data['id'])
 
+    def pubsub_event_filter(self, sub_id, event):
+        """Filter Pub/Sub events
+
+        Filter received Pub/Sub event using provided filter dictionary.
+        Return True if the event matches with the filter, otherwise False.
+        """
+        event_filter_status = True
+        filters = self._filters.get(sub_id)
+        if not filters:
+            return event_filter_status
+        for key, value in filters.items():
+            if key not in event.keys():
+                continue
+            if isinstance(value, dict):
+                for sub_key, sub_value in value.items():
+                    if sub_key not in event.get(key):
+                        continue
+                    if sub_value != event.get(key).get(sub_key):
+                        event_filter_status = False
+                        return event_filter_status
+            elif value != event[key]:
+                event_filter_status = False
+                break
+        return event_filter_status
+
+    def receive_node(self, sub_id):
+        """
+        Listen to all the events on 'node' channel and applies filter on it.
+        Return node if filter matches with the event, otherwise None.
+        """
+        path = '/'.join(['listen', str(sub_id)])
+        resp = self._get(path)
+        event = from_json(resp.json().get('data'))
+        node = self.get_node_from_event(event)
+        node['op'] = event.data['op']
+        event_filter_status = self.pubsub_event_filter(sub_id, node)
+
+        if event_filter_status is False:
+            return
+        return node
```

Comment:
I believe this should be a `while` loop to only return a Node object when it matches the filter.  From a caller point of view, it basically means `db.receive_node()` will always return a node matching the filters set when subscribing.

Reply:
We need to run `db.receive_node` inside `while` loop. It will return node object if filter matches, otherwise None.

```
while True:
            node = db.receive_node(sub_id)
            if node:
                print("Node received:", node)
```

Reply:
But what is the value of returning None to the caller when the method could just keep waiting until a matching node is received? We could just remove the `if node` part on the caller side if the method always returns a matching node. Errors would be raised as exceptions.

Reply:
Updated the PR.

Reply:
> Errors would be raised as exceptions.

What I meant is that we don't need to rely on the method to return `None` to propagate errors, as that can be one using exceptions.  So the `receive_node()` method will always return a valid node object, or throw an exception if an error occured.